### ISSUE TITLE
Configuring maximum number of threads at runtime

### ISSUE THREAD

**Issue by poulson** [flame/blis#61]
A user of Elemental has been running into strange performance issues when building on top of BLIS that seem to be due to the environment variable `OMP_NUM_THREADS` being set to one still leading to a large performance degradation when the number of independent uses of BLIS times the number of configured threads is larger than the number of cores on the machine.

While they have configured BLIS to use OpenMP with 16 threads, it is often preferred when using the library from within an MPI environment to be able to disable threading at runtime (or at least, decrease the number of threads).

What environment variables should be set to one to have the same effect as exporting `OMP_NUM_THREADS=1` for other BLAS libraries? I would humbly suggest either adding support for `OMP_NUM_THREADS` or adding the full list of variables (including `BLIS_IR_NT`) to the wiki that need to be configured to have the same effect.

**Comment by devinamatthews** [comment]
The list of environment variables is given in the [Wiki](https://github.com/flame/blis/wiki/Multithreading). These are read by BLIS at runtime, although I do not know if changing them during execution has any effect.

**Comment by poulson** [comment]
Thanks! In hindsight, that seems much more clear. So, to verify, a user needs to run:

    export BLIS_JC_NT=1
    export BLIS_IC_NT=1
    export BLIS_JR_NT=1
    export BLIS_IR_NT=1

to have the same effect as what `export OMP_NUM_THREADS=1` might otherwise imply? It might save a lot of user time if, when a user has defined `OMP_NUM_THREADS`, a check that the product of these environment variables exceeding `OMP_NUM_THREADS` fails that an appropriate fallback could be performed.

**Comment by devinamatthews** [comment]
One day...

**Comment by poulson** [comment]
In the mean time, should I recommend that users set `BLIS_JC_NT`, `BLIS_IC_NT`, `BLIS_JR_NT`, and `BLIS_IR_NT` to one?

**Comment by jeffhammond** [comment]
BLIS really shouldn't ignore `OMP_NUM_THREADS`.  That's evil.

In any case, I recommend supporting `BLI(S)_NUM_THREADS`, to align with e.g. `MKL_NUM_THREADS` and `MAD_NUM_THREADS` (in MADNESS).

**Comment by devinamatthews** [comment]
Yes.

**Comment by fgvanzee** [comment]
@poulson, @jeffhammond: I'm not convinced why BLIS should continue this `BLAH_NUM_THREADS` insanity. BLIS explicitly forces the user/developer to think about parallelism more deeply, and I think that is a good thing. Personally, I don't really care for using environment variables at all; instead, I prefer a runtime function call, but that is a minor quibble since the infrastructure for doing so is in place. And as soon as you have a runtime function, someone will suggest that we *also* support the environment variable method, and then we're more-or-less back in the same boat.

The basic point is this: Knowing the top-line number of threads isn't enough in order to reconstruct a decent parallelization scheme. Furthermore, not only is the top-line number not enough, I would venture to say it is borderline useless without additional information, unless we're just going to decree that it defaults to some arbitrary parallelization strategy (eg: always parallelize the 2nd loop around the micro-kernel, and only that loop).

About the only useful thing I can think of with such a hypothetical `BLIS_NUM_THREADS` is as a way to disable parallelism. Or was that the whole point of this discussion (just to be able to concisely disable threading)?

**Comment by poulson** [comment]
@fgvanzee Yes, the ability to easily disable parallelism was exactly the point. But, more generally, the ability to cleanly scale back the number of available cores would be of use.

**Comment by fgvanzee** [comment]
@poulson: I think there is an NSF grant proposal in there: How to better specify and control parallelism in libraries, and communicate and parameterize parallelism back and forth between library and application (or library-caller).

In fact, I think we may have put it into our pending NSF grant. I'll have to check with Robert.

**Comment by devinamatthews** [comment]
One thing you could do is to tabulate for #threads = 1...$MAX_THREADS 
what the optimal strategy is and then just look it up. For #threads > 
$MAX_THREADS you can just do whatever because it will suck anyways. 
However, you might also want to adjust for the shape of the matrices at 
runtime.

**Comment by poulson** [comment]
I honestly think this is just five lines of code in a `bli_init` function (assuming it exists). I will just submit a PR.

**Comment by poulson** [comment]
It seems that, without a bit of refactoring, this needs to be done in several files, but this PR handles Gemm: https://github.com/flame/blis/pull/62

**Comment by jeffhammond** [comment]
@fgvanzee Sorry, but Berkeley already did this with [Lithe](http://lithe.eecs.berkeley.edu/).

In any case, `BLAH_NUM_THREADS` is how application developers expect to be able to do things.  They are also accustomed to runtime calls like `{omp,mkl}_set_num_threads(int n)`.  And there is use beyond just `BLAH_NUM_THREADS=1`, because some of us like to call GEMM from Pthreads and not oversubscribe.

**Comment by jeffhammond** [comment]
@poulson If the only way to disable parallelism is `bli_init`, then all of us Fortran application users who want to use BLIS are hosed.  I'm not saying it's your job to implement other methods, but I'm not sure we should close this issue just because your PR is merged...

**Comment by poulson** [comment]
@jeffhammond The changeset that I linked above directly modifies the Gemm threading routine and does not involve an init function. But an equivalent approach would need to be manually propagated to several other routines.

**Comment by fgvanzee** [comment]
@jeffhammond I am congnizant of the "application-level `gemm()`" use case, however, I don't see how `BLAH_NUM_THREADS=n` specifically addresses that case better than a runtime function call. Could you elaborate?

**Comment by jeffhammond** [comment]
@fgvanzee I can put environment variables in a run script.  A runtime function call requires source changes.  Do you expect me to change the source code and build system of NWChem in order to allow users to link against BLIS compiled with thread support?  No other BLAS library requires that.

**Comment by fgvanzee** [comment]
@jeffhammond I understand what you are referring to now. However, this has nothing to do with `BLAH_NUM_THREADS=n` since I've already made the case that a single variable is generally not sufficient to communicate an effective threading strategy to BLIS.

**Comment by jeffhammond** [comment]
@fgvanzee Do you think that it is impossible to prescribe the four different thread variables from the total thread count for each platform?  As I recall, there are pretty simple rules based upon the cache structure for reducing the places where threading should be used.

In any case, my prior comment was addressing the question of environment variable versus runtime call _for providing the same information_, and meant to debate one variable versus four.

**Comment by fgvanzee** [comment]
@jeffhammond That *might* be reasonable for a single value of `_NUM_THREADS`, such as the maximum that would make sense on a given hardware configuration. But you just gave the example of using a value that is less than that maximum. Enumerating the strategy for every value from 1 to n_max seems excessive. Plus, this doesn't even account for the fact that you may have different numbers of cores on your actual hardware configuration for any given architecture (e.g. the Haswell under my desk has fewer cores than server-based Haswells).

Bottom line: This is not a can of worms I want to open.

As for the environment variable vs. runtime function aspect, I don't have much of a dog in that fight. If we can support both, everyone is happy and we can all drink beer together at the end of the day.

**Comment by jeffhammond** [comment]
@fgvanzee Sure, but I expect the user to set `BLIS_NUM_THREADS` to the number of cores in their system, or perhaps 2x that number, so you need only worry about the values of the four thread levels that can be obtained from the prime factors of that value.

Anyways, I think this is further evidence that a BLAS _product_ derived from BLIS needs some additional infrastructure to align the knobs to the specific hardware and application usage.  We know that MKL solves these problems, but their budget is only _slightly_ larger 😉

**Comment by magania** [comment]
I think the following heuristic is good enough: Translate _NUM_THREADS=N  to BLIS_JC_NT=P, BLIS_IC_NT=Q, BLIS_JR_NT=1, BLIS_IR_NT=1  where N = P*Q  and Q > P,  If more than one factorization exist chose the one that makes Q - P smaller.

**Comment by tlrmchlsmth** [comment]
This could work OK, except I would rather set BLIS_JR_NT to P rather than the JC loop in order to avoid having multiple blocks of B in the L3 cache

**Comment by devinamatthews** [comment]
This has been fixed by #109.
